server.js

```diff
@@ -38,168 +43,178 @@ const tracks = {
-	let goodTag = (tag === 'nightly' || tag.startsWith('v'));
+	if (!goodTag) {
+		throw new Error(`Invalid tag: ${tag}`);
+	}
 
-	let out;
 	console.log(`Pushing commit: ${commit} (tag: ${tag}/${goodTag})`);
 
-	if (goodTag) {
-		request.get({headers: { 'User-Agent': githubRepo }, url: `https://raw.githubusercontent.com/${githubRepo}/${commit}/util/src/misc.rs`}, function (error, response, body) {
-			if (error) throw error;
-			let branch = body.match(`const THIS_TRACK. ..static str = "([a-z]*)";`)[1];
-			let track = tracks[branch] ? branch : 'testing';
-			console.log(`Track: ${branch} => ${track} (${tracks[track]}) [enabled: ${enabledTracks[track]}]`);
-
-			if (enabledTracks[track]) {
-				request.get({headers: { 'User-Agent': githubRepo }, url: `https://raw.githubusercontent.com/${githubRepo}/${commit}/ethcore/src/ethereum/mod.rs`}, function (error, response, body) {
-					if (error) throw error;
-					let pattern = `pub const FORK_SUPPORTED_${network.toUpperCase()}: u64 = (\\d+);`;
-					let m = body.match(pattern);
-					if (m === null) {
-						console.log(`Unable to detect supported fork with pattern: ${pattern}.`);
-						return;
-					}
-					let forkSupported = m[1];
-
-					out = `RELEASE: ${commit}/${track}/${branch}/${forkSupported}`;
-					console.log(`Fork supported: ${forkSupported}`);
-
-					request.get({headers: { 'User-Agent': githubRepo }, url: `https://raw.githubusercontent.com/${githubRepo}/${commit}/Cargo.toml`}, function (error, response, body) {
-						if (error) throw error;
-						let version = body.match(/version = "([0-9]+)\.([0-9]+)\.([0-9]+)"/).slice(1);
-						let semver = +version[0] * 65536 + +version[1] * 256 + +version[2];
-
-						console.log(`Version: ${version.join('.')} = ${semver}`);
-
-						api.parity.registryAddress().then(a => {
-							console.log(`Registry address: ${a}`);
-							let registry = api.newContract(RegistrarABI, a);
-							return registry.instance.getAddress.call(
-								{},
-								[operationsContract, 'A']
-							);
-						}).then(a => {
-							console.log(`Parity operations address: ${a}`);
-							console.log(`Registering release: 0x000000000000000000000000${commit}, ${forkSupported}, ${tracks[track]}, ${semver}, ${isCritical}`);
-							// Should be this...
-							// api.newContract(OperationsABI, a).instance.addRelease.postTransaction({from: account.address}, [`0x000000000000000000000000${commit}`, forkSupported, tracks[track], semver, isCritical])
-							// ...but will have to be this for now...
-							return sendTransaction(OperationsABI, a, 'addRelease', [`0x000000000000000000000000${commit}`, forkSupported, tracks[track], semver, isCritical]);
-						}).then(h => {
-							console.log(`Transaction sent with hash: ${h}`);
-						});
-					});
-				});
-			}
-		});
+	const miscBody = await fetchFile(commit, '/util/src/misc.rs');
+	const branch = miscBody.match(`const THIS_TRACK. ..static str = "([a-z]*)";`)[1];
```

Comment:
Could the match fail?

Reply:
Indeed it fails if the file is not there (that's actually current behaviour of the server).